pallets/lbp/src/lib.rs

```diff
@@ -0,0 +1,1017 @@
+		) -> DispatchResultWithPostInfo {
+			T::CreatePoolOrigin::ensure_origin(origin)?;
+
+			ensure!(
+				!asset_a.amount.is_zero() || !asset_b.amount.is_zero(),
+				Error::<T>::CannotCreatePoolWithZeroLiquidity
+			);
+
+			ensure!(asset_a.id != asset_b.id, Error::<T>::CannotCreatePoolWithSameAssets);
+
+			let asset_pair = AssetPair {
+				asset_in: asset_a.id,
+				asset_out: asset_b.id,
+			};
+
+			ensure!(!Self::exists(asset_pair), Error::<T>::PoolAlreadyExists);
+
+			ensure!(
+				T::MultiCurrency::free_balance(asset_a.id, &pool_owner) >= asset_a.amount,
+				Error::<T>::InsufficientAssetBalance
+			);
+
+			ensure!(
+				T::MultiCurrency::free_balance(asset_b.id, &pool_owner) >= asset_b.amount,
+				Error::<T>::InsufficientAssetBalance
+			);
+
+			let pool_data = Pool::new(asset_a, asset_b, sale_duration, weight_curve, pausable);
+			Self::validate_pool_data(&pool_data)?;
+
+			let pool_id = Self::get_pair_id(asset_pair);
+
+			let deposit = T::PoolDeposit::get();
+
+			T::MultiCurrency::reserve(T::NativeAssetId::get(), &pool_owner, deposit)?;
+			<PoolDeposit<T>>::insert(&pool_id, &deposit);
+
+			<PoolOwner<T>>::insert(&pool_id, &pool_owner);
+			<PoolAssets<T>>::insert(&pool_id, &(asset_a.id, asset_b.id));
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+
+			T::MultiCurrency::transfer(asset_a.id, &pool_owner, &pool_id, asset_a.amount)?;
+			T::MultiCurrency::transfer(asset_b.id, &pool_owner, &pool_id, asset_b.amount)?;
+
+			<PoolBalances<T>>::insert(&pool_id, &(asset_a.amount, asset_b.amount));
+
+			Self::deposit_event(Event::PoolCreated(pool_owner, asset_a.id, asset_b.id, asset_a.amount, asset_b.amount));
+
+			Ok(().into())
+		}
+
+		#[pallet::weight(<T as Config>::WeightInfo::update_pool_data())]
+		#[transactional]
+		pub fn update_pool_data(
+			origin: OriginFor<T>,
+			pool_id: PoolId<T>,
+			start: Option<T::BlockNumber>,
+			end: Option<T::BlockNumber>,
+			initial_weights: Option<((AssetId, LBPWeight), (AssetId, LBPWeight))>,
+			final_weights: Option<((AssetId, LBPWeight), (AssetId, LBPWeight))>,
+		) -> DispatchResultWithPostInfo {
+			let who = ensure_signed(origin)?;
+
+			ensure!(<PoolOwner<T>>::contains_key(&pool_id), Error::<T>::PoolNotFound);
+
+			Self::ensure_pool_ownership(&who, &pool_id)?;
+
+			let mut pool_data = Self::pool_data(&pool_id);
+
+			let now = <frame_system::Pallet<T>>::block_number();
+			ensure!(
+				pool_data.start == Zero::zero() || now < pool_data.start,
+				Error::<T>::SaleStarted
+			);
+
+			if let Some(new_start) = start {
+				pool_data.start = new_start;
+			}
+
+			if let Some(new_end) = end {
+				pool_data.end = new_end;
+			}
+
+			if let Some((w1, w2)) = initial_weights {
+				pool_data.initial_weights = (w1, w2);
+			}
+
+			if let Some((w1, w2)) = final_weights {
+				pool_data.final_weights = (w1, w2);
+			}
+
+			Self::validate_pool_data(&pool_data)?;
+
+			<PoolData<T>>::insert(&pool_id, &pool_data);
+			Self::deposit_event(Event::PoolUpdated(who, pool_id));
```

Comment:
```suggestion
			<PoolData<T>>::try_mutate(pool_id.clone(), |pool| -> DispatchResult {
				let now = <frame_system::Pallet<T>>::block_number();

				ensure!(
					pool.start == Zero::zero() || now < pool.start,
					Error::<T>::SaleStarted
				);

				pool.start = start.unwrap_or(pool.start);
				pool.end = end.unwrap_or(pool.end);
				pool.initial_weights = initial_weights.unwrap_or(pool.initial_weights);
				pool.final_weights = final_weights.unwrap_or(pool.final_weights);

				Self::validate_pool_data(&pool)?;
				Self::deposit_event(Event::PoolUpdated(who, pool_id));

				Ok(())

			})?;

```

Something like this.

But the owner check is missing. Hover it is not tested! because all tests are passing with this change.!